### DISCUSSION TITLE
on demand animation

### DISCUSSION THREAD

**Discussion by SET001** [jcornaz/benimator#39] [Q&A]
How can I have animation to run once on demand? I need to show shot splash animation each time player does a shot. I see 2 options here:
1. on each shot to add a new entity with SpriteSheetBundle and animation and then track Play component removal and remove whole this entity. But this way I will remove all entities that jsut lost Play component, not only shot animation entity.
2. create entity beforehand with SpriteSheetBundle and animation but without Play component and with `visibility: Visibility{is_visible: false},`. And then on each shoot to add `Play` component in it and set up visibility to true. The problem here is how do I differ shot animation entity from other animation entities?

**Comment by jcornaz** [comment]
> on each shot to add a new entity with SpriteSheetBundle and animation and then track Play component removal and remove whole this entity. But this way I will remove all entities that jsut lost Play component, not only shot animation entity.

You may ensure that a special `Splash` component is present before removing the entity.

> create entity beforehand with SpriteSheetBundle and animation but without Play component and with visibility: Visibility{is_visible: false},. And then on each shoot to add Play component in it and set up visibility to true. The problem here is how do I differ shot animation entity from other animation entities?

I don't think I understand well. Why *beforehand*? You could spawn the splash animation when you need it. When it comes to differentiating shot animation from other entities, it's the same as how to differentiate any entity. You may use a "tag" component to recognize the purpose of the entity.


Lastly, be aware that you can also reset the animation by calling [`reset` on the `SpriteSheetAnimationState`](https://docs.rs/benimator/latest/benimator/struct.SpriteSheetAnimationState.html#method.reset). So you may re-use the same entity, add/remove the play component and reset the animation each time.

Personally, I think I would go the route of spawning a new "splash" animation entity, and despawn it when it is done by detecting when the `Play` component is removed. I would use a `Splash` unit struct to tag the splash entities in order to make sure that I despawn only the splash. Example:


```rust
fn despawn_splashes(mut cmds: Commands, splashes: Query<(), With<Splash>>, ended: RemovedComponent<Play>) {
  for entity in ended.iter().filter(|e| splashes.get(e).is_ok()) {
    cmds.entity(entity).despawn();
  }
}
```

**Comment by SET001** [comment]
`I don't think I understand well. Why beforehand?`

I mean to add all animation entities when the parent entity is created (Player may have multiple animations - shot, engine fire, and some effects) but without `Play` component and with `visibility` set to false. Then when player click to shot, add Play component and set visibility to true.

Adding tag components to select the exact animation amongst others, would definitely work for me. Thanks.
